perfs low hanging fruits

mpenet opened the pull request:
It's a work in progress but I found a few areas where things can be made faster and more gc friendly. 

I didn't measure the overall impact (yet), but I did bench the changes separately to make sure they are indeed improvements. 

I am still wrapping my head around the implementation to see if we can do more extremes changes (other than changing the surface/impl of some operation).

mpenet committed:
slightly faster order-on-edn-types

mpenet committed:
faster satisfies?

mpenet committed:
faster nth-of-set

mpenet committed:
add nth-index-node-keys

mpenet committed:
use nth-index-node-keys

mpenet committed:
avoid calling sorted-map-by twice

mpenet committed:
kill doubled lookups

mpenet committed:
decent jvm-opts for benchmarking

mpenet commented:
Running some benchmarks it seems to not bring that much (or anything) or maybe I failed to understand how to run/read the benchmarks. 

In isolation the changes are effective but in the grand scheme of things: not really. I guess it's time to dig into the logic of it all now and use a profiler instead of shooting blind.

whilo commented:
I think both David and me have already spent many hours with profilers on the codebase. Any speed-up is welcome, of course!

mpenet committed:
undo eduction but use pop instead of butlast on children vector

mpenet committed:
"inline" assoc-in

mpenet committed:
into [] + xform seems faster than mapv (!)

mpenet committed:
we only care about 1 node so don't run last-key on all of them up to point

mpenet committed:
use transients when possible

mpenet commented:
ok, getting somewhere. I get improvements. I am relying on the current benchmark suite but I am tempted to make a criterium or JMH based thing to get more precise results as it seems I get sometimes wildly different results for the same code. 

If you could give a look to the changes I'd appreciate. I will try a bit more of these kind of optimizations then I ll re-organize this mess of commits in something you could merge (hopefully).

mpenet committed:
make sure we have the same jvm flags as master for benchmarking

mpenet commented:
related: https://github.com/replikativ/hasch/pull/6

mpenet committed:
clojure :or destructuring will eval the values even when not
necessary/used

see: https://dev.clojure.org/jira/browse/CLJ-1676

mpenet committed:
kill useless local

mpenet committed:
reuse intermediary val for (pop path) when we can (every cycle counts)

mpenet committed:
allow control over rnd seed

whilo commented:
Looks like nice work so far. I would love to get a sense for the speed-ups you are seeing.

mpenet committed:
make sure lein doesn't inject other flags

mpenet commented:
It varies quite a bit depending on the runs, but there seem to be always gains. In some cases this can be ~30% in other in the single digits. I am only relying on the current bench task at the moment. 
I suspect this branch is also more gc friendly overall. Also keep in mind I am running these benchmarks on my laptop, so I am not trusting the accuracy of these results as much as I would if I was running them in an isolated environment. I tried to do my best to limit the external "noise" but you never know. 

I put 2 spreadsheets on google docs to compare: 

* [spreadsheet for perf branch](https://docs.google.com/spreadsheets/d/1uFl3xldnhs0visXEwfjRt_vJrH8tzmpj9mjMGJlmvN4/edit?usp=sharing)
* [spreadsheet for master](https://docs.google.com/spreadsheets/d/1zmi8W0oyk-bZs467eDXBvMUPiLs_ur-TcfsHu43FbIA/edit?usp=sharing) 

The scales should be identical between the 2 spreadsheets graphs, opening both in separate tabs and going back and forth gives a somewhat decent view of the gains.  

master is modified to use same bench task and same jvm args (actual branch is here https://github.com/mpenet/hitchhiker-tree/tree/master-p)


This is done via: `lein bench perf_diff_experiment -b 10 -r 1 -- -b 20 -r 1 -- -b 40 -r 1 -- -b 80 -r 1 -- -b 160 -r 1 -- -b 320 -r 1 -- -b 640 -r 1`

mpenet commented:
Just noticed I have failures in hitchhiker.tree.messaging-test :(

mpenet committed:
actually remove is-first? check since we refactored it lower

mpenet commented:
fixed

mpenet committed:
add (last-key node) caching

mpenet commented:
more gains with last-key caching on latest commit. I added the link to the spreadsheet in the original comment with the other two. 

I guess I could do this in a more functional way now that I think of it. I can give it a try tomorrow.

whilo commented:
Very nice work. I will have not time to look at it until Monday, but will give feedback afterwards.

whilo commented:
@mpenet How do you get along with dual backend (core.async vs. sync) btw.? We are worried whether this is pulling off potential contributors. On the other hand it allows to support JavaScript, which is kind of a big benefit.

mpenet commented:
I like it. I usually worry about misuse or overuse of core.async but in that case it fits nicely. It has to be something optional tho, so having both sync/async interfaces makes sense imho, even if that means just building the sync api on top of the async one.
There's one case where I have yet to make my mind is with superv.async in datahike, but it's probably because I am just discovering it. I have decent experience with erlang/otp so I guess this will click but it might be off-putting for potential contributors.
I usually see/did error propagation with "control" channels (could be even simple promise-chans) between contexts, that can signal errors that way but I suspect that superv.async prolly does something similar + some plumbing for dealing with exceptions. But as I said I am not really familiar with its internals (yet).

mpenet committed:
we already have a ref to :children, avoid a map lookup

mpenet commented:
latest one: https://docs.google.com/spreadsheets/d/12NF3sO9JdTonfYCGMM5MV1SCeRAz3V5XDgIaNRU9--A/edit?usp=sharing

I tried to replace clojure.core/split-with with an equivalent transducer so that we can do a single pass over index-node :op-buf when we have to split (the core version will do take-while+drop-while) and avoid going from lazy-seqs to vecs later too. 
I went also further by trying to add the sort-by pass in the transducer chain (via cgrand/xform xf/sort-by) but no noticeable improvement either. 
I *might* very well be messing up something while doing it which would cause the lack of gains but I didn't have the time to dig further. 
 
I didn't commit it, the results are not really conclusive but it's here to play with in case. It mostly matches the perf of the cached version with more stability in trial5, could be because of less gc pressure, or just throttling on my laptop, hard to say. 

The xform is here if you want to play with it : 

```
(defn x-split-with [f]
  (fn [rf]
    (let [;; we could go nuts and use a native array but for now this
          ;; will do and this works with cljs too
          state (volatile! (transient []))
          pivot? (volatile! false)]
      (fn
        ([] (rf))
        ([result]
         (rf (rf result (persistent! @state))))
        ([result input]
         (let [acc @state]
           (if @pivot?
             (do (vswap! state conj! input)
                 result)
             (if (f input)
               (do (vswap! state conj! input)
                   result)
               (do
                 (vreset! pivot? true)
                 (vreset! state (transient [input]))
                 (rf result (persistent! acc)))))))))))

```

and later:


```
[...]
          [left-buf right-buf] (into [] (comp (xf/sort-by :key compare) ;; via cgrand/xform
                                              (x-split-with #(not (pos? (compare (:key %) median)))))
                                       op-buf)
```
or without xform: 
```
[...]
          [left-buf right-buf] (into [] (x-split-with #(not (pos? (compare (:key %) median)))))
                                       (sort-by :key compare op-buf))
```

mpenet committed:
much faster/cheaper index-node split-node

mpenet commented:
I updated the spreadsheet links with the last modifications.

mpenet committed:
missing negation

mpenet committed:
slightly faster/more efficient data-node split

mpenet committed:
make sure we can still serialize for konserve

Having to mess with the removal/resetting of "impl" fields is a bit a
dirty buisness. Makes one wonder if we should think about handling
this in a more generic way

mpenet commented:
While fixing the tests for konserve I was thinking maybe it would be nicer not to have to do this kind of things (nilify some fields like storage-addr/last-key-cache) at "storage" level and have an intermediary serializable representation of various nodes we can generate from .core , maybe something like a tagged hash-map.   

* a new protocol + functions for converting nodes to something more transparent than records -> a simple tagged hashmap `{::hh-tree.node-type :index-node + <rest of the fields> }` 
* and from handlers in storage we just call these without having to worry about extra fields like stuff related to caching or other internal bits.

... or we can just have a function (in core) that sets these fields to "nil" and be done with it. I find it slightly ugly/inefficient to have to store nil fields tho.

mpenet committed:
fix bench task

whilo commented:
> There's one case where I have yet to make my mind is with superv.async in datahike, but it's probably because I am just discovering it. I have decent experience with erlang/otp so I guess this will click but it might be off-putting for potential contributors.

I have decided not to use superv.async in the end, because the overhead for the book-keeping seemed to add up in my experiments (which probably is not valid anymore with all the optimizations :) ) and we do not really have concurrency here, so I just use go-try to propagate IO errors with stack traces. 

> I usually see/did error propagation with "control" channels (could be even simple promise-chans) between contexts, that can signal errors that way but I suspect that superv.async prolly does something similar + some plumbing for dealing with exceptions. But as I said I am not really familiar with its internals (yet).

Yes. It actually also terminates concurrent go-routines and detects errors that are not taken care off (not taken off a channel). I would say it is a fairly good take to make core.async robust in an Erlang fashion, but maybe also a bit of a stretch. I would really love to hear your feedback, because I have only done tutorial code about Erlang and studied its error handling.

whilo commented:
> While fixing the tests for konserve I was thinking maybe it would be nicer not to have to do this kind of things (nilify some fields like storage-addr/last-key-cache) at "storage" level and have an intermediary serializable representation of various nodes we can generate from .core , maybe something like a tagged hash-map.
> 
>     * a new protocol + functions for converting nodes to something more transparent than records -> a simple tagged hashmap `{::hh-tree.node-type :index-node + <rest of the fields> }`
> 
>     * and from handlers in storage we just call these without having to worry about extra fields like stuff related to caching or other internal bits.
> 
> 
> ... or we can just have a function (in core) that sets these fields to "nil" and be done with it. I find it slightly ugly/inefficient to have to store nil fields tho.

Yes, totally agree. Lets use a protocol for that, I think. The nils are there to keep the record representation intact, which is kind of stupid. I think we can do this after this PR.

mpenet commented:
Imho let it crash is quite often mentioned with incorrect interpretations, I found that in practice you end up being extra defensive against failures and "let it crash" only as last resort, retries/restarts are far from free and supervisors not magical. Sure, you rarely use exceptions in erlang but you shield yourself from errors with extensive pattern matching, dialyzer and other means. I have never found the need to design that way systems with clojure/java. Also in erlang processes are quite cheap, contained, with a memory model that allows to design last resort failure mode that way but it's not very well suited for the jvm in general imho.

I have used core.async in a few projects and I tend to convey errors as values (tagged or raw exceptions instances) simply, and let the user deal with it as they wish.

You can for instance create channels with a xform that will turn messages into something like `[:ok <value>]` `[:error value]` and reuse that xform everywhere: (async/chan ... (map as-result)), then it's a simple pattern match with `case` away in your code, which is quite decent for a very small cost. Another way to do it is to return (and potentially take as input) a "control" chan that will contain just errors and values that triggered their occurrence they you can handle these downstream with alt!(!) instead of case as before, but I prefer the other approach overall.

I try also to avoid using return channel from go blocks directly and prefer to create function that can take and return the input channel(s) instead, this way the user can specify type of chan, potential xform(s) at that level, buffering etc. This also allows some optimizations like using promise-chan instead of a full blown chan as return value, which opens the door to multiple takers without lots of copying etc etc. In short I use go blocks to "coordinate/manage" operations on channels but I prefer to control what's returned myself instead of deferring to go blocks channel return. 

But I digress, this is another broad subject we can talk about on slack maybe.

whilo commented on src/hitchhiker/tree/core.cljc:
```diff
@@ -394,31 +436,62 @@ throwable error."
-    (->Split (data-node cfg (into (sorted-map-by compare) (take (:data-b cfg)) children))
-             (data-node cfg (into (sorted-map-by compare) (drop (:data-b cfg)) children))
-             (nth-of-set children (dec (:data-b cfg)))))
+    (let [data-b (:data-b cfg)]
```
Do we do this because of the intermediary lazy sequence allocation?

mpenet replied:
among other things yes, it saves allocating 2 lazy seqs and iterating over the children twice as well.

whilo replied:
Right.

whilo commented on src/hitchhiker/tree/core.cljc:
```diff
@@ -431,7 +504,35 @@ throwable error."
+(def satisfies?
+  ;; FIXME for now `(satisfies? IResolve node)` is crippled,
+  ;; implementers have no cache see:
+  ;; https://dev.clojure.org/jira/browse/CLJ-1814.  So we just check
```
Interesting, didn't know this. Thanks!

whilo commented on src/hitchhiker/tree/core.cljc:
```diff
@@ -652,93 +748,97 @@ throwable error."
-                         (assoc-in [:children index] node)
-                         (dirty!))
-                     (pop (pop path))))))))))
+   (let [path (<? (lookup-path tree key))
```
The indentation here is a bit unfortunate to see what has actually changed. It looks it is only cosmetic changes.

mpenet replied:
Sorry about that, my emacs config re-indents code aggressively. 
Imho it would be good to do a full pass with clj-fmt on the codebase, but this is out of scope here.

mpenet replied:
Actually `init-path` is new, we had multiple calls to (pop path) in the same iteration, ex line 666 and 676/681

whilo replied:
Ok, thanks.

whilo commented on src/hitchhiker/tree/core.cljc:
```diff
@@ -652,93 +748,97 @@ throwable error."
-                                        cfg)
-                           (pop (pop path))))
-                  (recur (->IndexNode (catvec (conj old-left-children merged)
+   (let [path (<? (lookup-path tree key)) ; don't care about the found key or its index
```
Same here, the indentation change makes it difficult to see what has actually changed.

mpenet replied:
same as before, avoiding to call (pop path) multiple times

whilo replied:
Right.

whilo commented on src/hitchhiker/tree/core.cljc:
```diff
@@ -750,6 +850,7 @@ throwable error."
 (defrecord TestingAddr [last-key node]
   IResolve
   (dirty? [this] false)
+  (dirty! [this] this)
```
This is a good idea!

whilo commented on src/hitchhiker/tree/core.cljc:
```diff
@@ -182,21 +183,42 @@ throwable error."
+(defprotocol OrderOnEDNTypes
+  (order-on-edn-types [t]))
+
+(extend-protocol OrderOnEDNTypes
```
Does this protocol also work for cljs atm.? I am not sure whether I like people changing the order of the core edn types, as it can introduce weird problems later if indices are distributed and synchronized, but it is good to make it optionally extensible to other types, e.g. records. I guess we can hope that nobody is extending the protocol on application side overwriting core types. (And we should directly add UUID support as it is part of edn (I can also do it).)

mpenet replied:
I have no idea, I assumed it would, but I will double check now.

mpenet replied:
should be fixed by 20ed4c2

whilo commented on env/profiling/hitchhiker/bench.clj:
```diff
@@ -10,9 +10,10 @@
-  []
-  [{:name "in-order" :data (range)}
-   {:name "random" :data (repeatedly rand)}])
+  [seed]
```
This is good.

whilo commented on src/hitchhiker/tree/core.cljc:
```diff
@@ -275,43 +297,62 @@ throwable error."
   (last-key [this]
-    ;;TODO should optimize by caching to reduce IOps (can use monad)
-    (last-key (peek children)))
+    (<-cache *last-key-cache
```
Nice. I guess this is the reason for most of the speedups? The benchmarks test write operations mostly, but we should see large speedups on reads by this as well.

mpenet replied:
Yes it contributes quite a bit.

whilo replied:
Good, very nice.

whilo reviewed (commented):
Hey, this overall looks very good. I would just like to get some clarifications and then merge it.

mpenet committed:
make order-on-edn-types cljs compatible

mpenet commented:
Weird, I will try to have a look as well.

mpenet commented:
It seems it could be related to https://dev.clojure.org/jira/browse/CRRBV-15, simply bumping the dependencies could fix it. I will give it a try and report back here. 

My bad I didn't test the cljs build, I should have really

mpenet committed:
fix cljs compilation

* bump core.rrb-vector to 0.0.14

* use object instead of :default in order edn types protocol

* make satisfies? work with both clj/cljs

mpenet commented:
seems ok now: 

```shell
$ lein cljsbuild once
Compiling ClojureScript...
WARNING: pos-int? already refers to: #'clojure.core/pos-int? in namespace: taoensso.encore, being replaced by: #'taoensso.encore/pos-int?
WARNING: bytes? already refers to: #'clojure.core/bytes? in namespace: taoensso.encore, being replaced by: #'taoensso.encore/bytes?
Compiling ["resources/public/js/core.js"] from ["src"]...
Successfully compiled ["resources/public/js/core.js"] in 3.239 seconds.
Compiling ["target/test.js"] from ["src" "test"]...
Successfully compiled ["target/test.js"] in 4.923 seconds.
```

whilo commented:
This is annoying, somehow `node test_node.js` does not run the tests anymore... Investigating. Ideally they should be invoked with the clj test, but we have not had the time yet to fix that.

whilo commented:
Ok, it is the `<-cache` macro that somehow expands wrong in cljs. I have made some progress, but it is not working yet.

mpenet commented:
Weird. We can probably just make it a function, we actually have little
reason to have it as a macro.

On Sun, Feb 17, 2019, 10:37 Christian Weilbach <notifications@github.com
wrote:

> Ok, it is the <-cache macro that somehow expands wrong in cljs. I have
> made some progress, but it is not working yet.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/replikativ/hitchhiker-tree/pull/3#issuecomment-464434511>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGflgwiagpv-LnTv3CETobp4_yMTv-5ks5vOSLlgaJpZM4aSqJA>
> .
>

whilo commented:
Ok, that makes sense.

whilo commented:
Ok, I got it figured out and checked the tests in the REPL, but the test runner for node is not working anymore.

whilo commented:
A problem besides the macro expansion was that `identical?` does not work in cljs for keywords, because they are not interned, it seems. I am pushing this fix now and you can take a look at it. I would appreciate if we can fix the node cljs runner. If you can do it, I would appreciate it, otherwise I will try to do so towards the end of the week. Thanks a lot for putting the effort in, I am loving it! I hope we can press on from here and make the codebase both more accessible and clean as well as more useful for datahike et al.

mpenet commented:
:champagne: 

Interesting: https://cljs.github.io/api/cljs.core/keyword-identicalQMARK 
I never knew about this difference.

About node cljs runner I will try, but I have a busy (short) week unfortunately.

whilo commented:
Ok. I am also fairly busy this week.
